### ISSUE TITLE
Test the recovered IEC61850 model using ModelDesign Compiler (mdc.exe) Version: 3.0

### ISSUE THREAD

**Issue by mpostol** [mpostol/UA-Nodeset#98]
## Describe the bug

Partially it contributes to mertyusaatag/UA-Nodeset#14
Partially it contributes to mpostol/OPC-UA-OOI#653.

**Comment by eoursel** [comment]
Hi @mpostol 

I am doing some testing with the last model compiler (the last one on the OPCF Members repository) and IEC61850/ 

The problem is the definition of the reference type HasDataAttribute which is a subclass of HasComponent as described in the IEC61850-7-3 NodeSet. 

The generated code looks like 

```
<ObjectType SymbolicName="Prefix3:SPS" BaseType="Prefix3:IEC61850DOBaseObjectType">
    <References>
      <Reference>
        <ReferenceType>Prefix3:HasDataAttribute</ReferenceType>
        <TargetId>Prefix3:stVal</TargetId>
      </Reference>
      <Reference>
        <ReferenceType>Prefix3:HasDataAttribute</ReferenceType>
        <TargetId>Prefix3:q</TargetId>
      </Reference>
```
but the compiler fails to solve the TargetId's

```
PS C:\Users\eoursel\source\repos\mpostol\UA-Nodeset\IEC61850\ModelDesign> Opc.Ua.ModelCompiler.exe compile -d2 .\Opc.Ua.IEC61850-7-3.Model.xml -cg .\Opc.Ua.IEC61850-7-3.csv -o2 .\IEC61850-7-3 -version v104
Ignoring InvalidReference SPS => stVal. [The TargetId reference for node SPS is not valid: stVal.]
Ignoring InvalidReference SPS => q. [The TargetId reference for node SPS is not valid: q.]
Ignoring InvalidReference SPS => subEna. [The TargetId reference for node SPS is not valid: subEna.]
Ignoring InvalidReference SPS => subVal. [The TargetId reference for node SPS is not valid: subVal.]
Ignoring InvalidReference SPS => subQ. [The TargetId reference for node SPS is not valid: subQ.]
Ignoring InvalidReference SPS => subID. [The TargetId reference for node SPS is not valid: subID.]
Ignoring InvalidReference SPS => blkEna. [The TargetId reference for node SPS is not valid: blkEna.]
Ignoring InvalidReference SPS => d. [The TargetId reference for node SPS is not valid: d.]
Ignoring InvalidReference SPS => dU. [The TargetId reference for node SPS is not valid: dU.]
Ignoring InvalidReference SPS => cdcNs. [The TargetId reference for node SPS is not valid: cdcNs.]
Ignoring InvalidReference SPS => cdcName. [The TargetId reference for node SPS is not valid: cdcName.]
Ignoring InvalidReference SPS => dataNs. [The TargetId reference for node SPS is not valid: dataNs.]
Ignoring InvalidReference DPS => stVal. [The TargetId reference for node DPS is not valid: stVal.]
Ignoring InvalidReference DPS => q. [The TargetId reference for node DPS is not valid: q.]
Ignoring InvalidReference DPS => subEna. [The TargetId reference for node DPS is not valid: subEna.]
Ignoring InvalidReference DPS => subVal. [The TargetId reference for node DPS is not valid: subVal.]
Ignoring InvalidReference DPS => subQ. [The TargetId reference for node DPS is not valid: subQ.]
Ignoring InvalidReference DPS => subID. [The TargetId reference for node DPS is not valid: subID.]
Ignoring InvalidReference DPS => blkEna. [The TargetId reference for node DPS is not valid: blkEna.]
Ignoring InvalidReference DPS => d. [The TargetId reference for node DPS is not valid: d.]
Ignoring InvalidReference DPS => dU. [The TargetId reference for node DPS is not valid: dU.]


```

Obviously the generated code should use ObjectType -> Children -> Variable. But in this case the HasDataAttribute references will be changed to HasComponent. 

Looks like a conceptual issue. 

1. First the idea to subclass HasComponent buy using HasDataAttribute in the IEC61850 model is questionable. 
2. Second point, the <children> tag in the model compiler has some limitations because it is not possible to change the default HasComponent (or HasProperty) references to variables and properties.  This behavior should require a modification of the ModelDesign Schema as well  as a change in the model compiler source code.   

what is your point of view on the subject? 

Regards

**Comment by mpostol** [comment]
Now available is version [Rel 3.1.1](https://github.com/mpostol/UA-ModelCompiler/releases/tag/3.1.1)

**Comment by mpostol** [comment]
Hi @eoursel, Many thanks for the feedback. Recently I worked on this issue so I will try to answer your question.  You will find general information about the project stage/progress and direct answer to your question.

## General stage and progress of the project

Thi issue was reported as a result of the project [Information Model Validation (IMV)](https://github.com/users/mpostol/projects/6) conducted thanks to the Erasmus program (internship and two final projects). The Erasmus Programme is a European Union student exchange program established in 1987. Unfortunately, the program is over and not all issues have been resolved.

To be honest, now all my activity targets my research related to the Object-Oriented Internet but not OPC UA. Still, OPC UA is a reference point and foundation for my research. Of course, in case of topics overlapping my findings could be useful also for the OPC UA community.  Today I am closing work on this project so as the result this issue will also be closed after creating a new release to make a final checkpoint. My next step will be project [Stand-alone Address Space](https://github.com/mpostol?tab=projects&type=classic#:~:text=alone%20Address%20Space%20(-,SAS,-)) conducted as a Master Thesis by @MichalSzczekocki. I will try to address this and other issues during this project if applicable.

## HasDataAttribute - ModelCompiler error

The model IEC61850 has not been managed properly during the mentioned project for some reasons. It is a reason I reported this issue to:

1. test the recovered models against [mpostol/UA-ModelCompiler](https://github.com/mpostol/UA-ModelCompiler)
2. test the recovered models against ASMD and generate missing online documentation

Let me stress I am not using the ModelCompiler originated by OPC Foundation for many reasons. Instead, I am using forked from OPC Foundation ModelCompiler but it was modified. 

Therefore my results are different and it is hard to compare my [results gathered in the folder ModelDesign](https://github.com/mpostol/UA-Nodeset/tree/main/IEC61850/ModelDesign).

BTW, I am using the NuGet package [OOI.ModelCompiler](https://www.nuget.org/packages/OOI.ModelCompiler) as compiler core.  The main objective for defining the compiler part as the NuGet package is reusability. I am going to use the same package by [ASMD](https://github.com/mpostol/ASMD) and [Address Space Prototyping Tool (asp.exe)](https://github.com/mpostol/OPC-UA-OOI/tree/master/SemanticData/AddressSpaceComplianceTestTool#address-space-prototyping-tool-aspexe).

It is hard to compare results but it is easy to state that HasDataAttribute can be derived from HasComponent (it is valid) and the ModelCompiler must be able to handle this. Because the Children element doesn't allow defining the reference type we must use the Reference element to accomplish it. As far as I remember my discussion with @randy-armstrong always we can use Reference instead of Children in case the reference is not the default for the members. BTW, I started working on a description of the [Model Design - Domain-Specific Language](https://github.com/mpostol/UA-ModelCompiler/tree/main/Opc.Ua.ModelCompiler#model-design---domain-specific-language-). Hopefully there will be a point in time that I will return to polishing this documentation. My point is that this must be harmonized with the further development of the ModelCompiler.

Hopefully, it helps. Let me know how it works for you.

Best regards,
Mariusz

**Comment by eoursel** [comment]
Hi Mariusz  Honestly i am fighting to generate the C# code for IEC61850-7-3 and IEC61850-4.  I have decided to change the HasDataAttribute and HasDataObject with HasComponent which is relatively easy in the NodeSet file. The problem is that we must generate a reference AND the variable as the target of the reference. The problem is that in the ModelDesign syntax the variable must be included in a object by a HasComponent reference… chicken and eggs problem.  There are many others issues in IEC61850 Nodesets since they are defining many (several hundreds) fields inside enumerated data types. Unfortunately, the field names are using some special characters (space, - , parenthesis). The model compiler tries to generate  C# identfiers for field names. Obvisouly, the C# compiler do not like it.  After some manual fixes in the original nodeset IEC68150-7-3, the model compiler generates a correct C# code which can compiled by the C# compiler.  I am always trying to understand why the model compiler fails on IEC61850-7-4,  it is not able to find the included model design IEC61850-7-3. I initialized filePath XML attribute without success.  I am playing with the modelcompiler from OPC Foundation (the last one from OPCF_Members) but i am also trying with your mdc tool to compare the results.  It is late, i should be in vacation in theory.  Thanks for the long answer and the work you have done since the beginning on this subject. The ASP tool is great, really.  Good night.  Eric  From: MariuszSent: vendredi 15 juillet 2022 23:03To: mpostol/UA-NodesetCc: Eric OURSEL; MentionSubject: Re: [mpostol/UA-Nodeset] Test the recovered IEC61850 model using ModelDesign Compiler (mdc.exe) Version: 3.0 (Issue #98) Hi @eoursel, Many thanks for the feedback. Recently I worked on this issue so I will try to answer your question. You will find general information about the project stage/progress and direct answer to your question.General stage and progress of the projectThi issue was reported as a result of the project Information Model Validation (IMV) conducted thanks to the Erasmus program (internship and two final projects). The Erasmus Programme is a European Union student exchange program established in 1987. Unfortunately, the program is over and not all issues have been resolved.To be honest, now all my activity targets my research related to the Object-Oriented Internet but not OPC UA. Still, OPC UA is a reference point and foundation for my research. Of course, in case of topics overlapping my findings could be useful also for the OPC UA community. Today I am closing work on this project so as the result this issue will also be closed after creating a new release to make a final checkpoint. My next step will be project Stand-alone Address Space conducted as a Master Thesis by @MichalSzczekocki. I will try to address this and other issues during this project if applicable.HasDataAttribute - ModelCompiler errorThe model IEC61850 has not been managed properly during the mentioned project for some reasons. It is a reason I reported this issue to:test the recovered models against mpostol/UA-ModelCompilertest the recovered models against ASMD and generate missing online documentationLet me stress I am not using the ModelCompiler originated by OPC Foundation for many reasons. Instead, I am using forked from OPC Foundation ModelCompiler but it was modified.Therefore my results are different and it is hard to compare my results gathered in the folder ModelDesign.BTW, I am using the NuGet package OOI.ModelCompiler as compiler core. The main objective for defining the compiler part as the NuGet package is reusability. I am going to use the same package by ASMD and Address Space Prototyping Tool (asp.exe).It is hard to compare results but it is easy to state that HasDataAttribute can be derived from HasComponent (it is valid) and the ModelCompiler must be able to handle this. Because the Children element doesn't allow defining the reference type we must use the Reference element to accomplish it. As far as I remember my discussion with @randy-armstrong always we can use Reference instead of Children in case the reference is not the default for the members. BTW, I started working on a description of the Model Design - Domain-Specific Language. Hopefully there will be a point in time that I will return to polishing this documentation. My point is that this must be harmonized with the further development of the ModelCompiler.Hopefully, it helps. Let me know how it works for you.Best regards,Mariusz—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>
